core/src/Kokkos_ReductionIdentity.hpp

```diff
@@ -389,6 +391,35 @@ struct reduction_identity<long double> {
   constexpr static long double min() { return LDBL_MAX; }
 };
 
+namespace Impl {
+template <typename T>
+using reduction_identity_sum_t = decltype(reduction_identity<T>::sum());
+
+template <typename T>
+KOKKOS_FUNCTION constexpr T reduction_identity_sum_or_value_initialize() {
+  if constexpr (Kokkos::is_detected_convertible_v<T, reduction_identity_sum_t,
+                                                  T>)
+    return reduction_identity<T>::sum();
+  else if constexpr (Kokkos::is_detected_v<reduction_identity_sum_t, T>)
+    static_assert(Kokkos::is_detected_v<reduction_identity_sum_t, T>,
+                  "Incorrect return type for reduction_identity<T>::sum(); "
+                  "should be convertible to T");
+  else if constexpr (std::is_default_constructible_v<T>)
+    return T{};
+  else
+    static_assert(!std::is_default_constructible_v<T>,
+                  "Kokkos::reduction_identity<T>::sum() is not defined nor is "
+                  "T default constructible");
+}
+
+template <typename T>
+KOKKOS_FUNCTION constexpr T& reduction_identity_sum_or_value_initialize(T& t) {
```

Comment:
I'm taking a non-const reference because that is what `Sum<...>::init` does, and I want the type deduction.  I was returning a reference as I experimented with passing the result directly to `Sum`, as in

```c++
value_type value;
parallel_scan(..., Sum<value_type>(reduction_identity_sum_or_value_initialize(value)), ...);
```

but ultimately went in favor of using the same pattern as `init` for consistency.  I can get rid of the `return`, as it isn't being used anywhere.